common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java

```diff
@@ -117,17 +117,23 @@ private FetchShuffleBlocks createFetchShuffleBlocksMsg(
     boolean batchFetchEnabled = firstBlock.length == 5;
 
     HashMap<Long, ArrayList<Integer>> mapIdToReduceIds = new HashMap<>();
+    ArrayList<Long> orderedMapId = new ArrayList<>();
     for (String blockId : blockIds) {
       String[] blockIdParts = splitBlockId(blockId);
       if (Integer.parseInt(blockIdParts[1]) != shuffleId) {
         throw new IllegalArgumentException("Expected shuffleId=" + shuffleId +
           ", got:" + blockId);
       }
       long mapId = Long.parseLong(blockIdParts[2]);
+      assert(orderedMapId.isEmpty() || mapId >= orderedMapId.get(orderedMapId.size() - 1));
```

Comment:
This bug reveals that the block ids order is very important. I think it makes sense to add asserts to guarantee it.

Reply:
I actually don't think that the bug reveals that order in blockIds is important. Just in `OneForOneBlockFetcher`, it never expects the provided `blockIds` are in certain order. Also at this layer why should it expect that?
It is given an array of blockIds which may be in any order and would fetch them. 
IMO, the bug is just that the  member `blockIds` of OneForOneBlockFetcher is out of sync with the FetchShuffleBlocks.

Reply:
> It is given an array of blockIds which may be in any order and would fetch them.

I think it's ordered because of `MapOutputTracker.convertMapStatuses`.

Reply:
My point is that the order of blockIds has been irrelevant for `OneForOneBlockFetcher` so far. At this layer, it doesn't have any dependency on the order of block Ids so why to introduce that at this layer now?

Reply:
So to be clarified, the order of `blockIds` could be random, but the `orderedMapId` should be in the same order with `blockIds`. And I think this PR fixes the latter.

Reply:
> the order of blockIds could be random

Well, this PR is adding the validation where it checks the mapIds in `blockIds` are in increasing order.  So the order of `blockIds` cannot be random. 
`assert(orderedMapId.isEmpty() || mapId >= orderedMapId.get(orderedMapId.size() - 1));`

Reply:
Ah, that's a good point. I agree the assertion is not necessary. But if we assume the random `blockIds` here, the current fix won't work with duplicate `mapIds`. (we should be able to save duplicate `mapIds` in `FetchShuffleBlocks` to ensure the server side serves chunks in the same order with `blockIds` )

Reply:
The initial fix that was proposed was to create the member `blockIds` in OneForOneBlockFetcher in accordance with the FetchShuffleBlocks message which should work even if blockIds are random.

Reply:
I see. Just feel a little bit uncomfortable about "rebuild"..

Reply:
I think we should reduce randomness in different task attempt and try to ensure the consistence of different task attempt.